src/Lean/Elab/PreDefinition/PartialFixpoint/Main.lean

```diff
@@ -72,24 +75,42 @@ def mkMonoPProd (hmono₁ hmono₂ : Expr) : MetaM Expr := do
   mkAppOptM ``PProd.monotone_mk #[none, none, none, inst₁, inst₂, inst, none, none, hmono₁, hmono₂]
 
 def partialFixpoint (preDefs : Array PreDefinition) : TermElabM Unit := do
-  -- We expect all functions in the clique to have `partial_fixpoint` syntax
+  -- We expect all functions in the clique to have `partial_fixpoint` or `greatest_fixpoint` syntax
   let hints := preDefs.filterMap (·.termination.partialFixpoint?)
   assert! preDefs.size = hints.size
+  -- We check if any fixpoints were defined lattice-theoretically
+  let isLattice := hints.any (fun x => isLatticeTheoretic x.fixpointType)
+  -- If yes, then we expect all of them to be defined using lattice theory
+  if isLattice then
+    assert! preDefs.size = (hints.filter (fun x => isLatticeTheoretic x.fixpointType)).size
+
+  -- We check if the
   -- For every function of type `∀ x y, r x y`, an CCPO instance
   -- ∀ x y, CCPO (r x y), but crucially constructed using `instCCPOPi`
   let ccpoInsts ← preDefs.mapIdxM fun i preDef => withRef hints[i]!.ref do
     lambdaTelescope preDef.value fun xs _body => do
       let type ← instantiateForall preDef.type xs
       let inst ←
-        try
-          synthInstance (← mkAppM ``CCPO #[type])
-        catch _ =>
-          trace[Elab.definition.partialFixpoint] "No CCPO instance found for {preDef.declName}, trying inhabitation"
-          let msg := m!"failed to compile definition '{preDef.declName}' using `partial_fixpoint`"
-          let w ← mkInhabitantFor msg #[] preDef.type
-          let instNonempty ← mkAppM ``Nonempty.intro #[mkAppN w xs]
-          let classicalWitness ← mkAppOptM ``Classical.ofNonempty #[none, instNonempty]
-          mkAppOptM ``FlatOrder.instCCPO #[none, classicalWitness]
+        match hints[i]!.fixpointType with
+        | .greatestFixpoint =>
+          if !type.isProp then
+              throwError "`greatest_fixpoint` can be only used to define predicates"
+          else
+            mkAppOptM ``inst_coind_complete_lattice #[]
```

Comment:
No need to involve type inference; this is a plain constant without level parameters:
```suggestion
            pure (mkConst ``inst_coind_complete_lattice)
```